### PR TITLE
Profile API Memory with Live Server

### PR DESCRIPTION
**Pull request by enjoythecode**

### CODE REVIEW

File: test/http_memprof/http_memprof.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// this code is a utility to monitor an already running mixer instance, and such
+// does not run with regular tests, so disable deadcode warning on it.
+//nolint:deadcode
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:17]: Curious what gives "deadcode" lint? This folder is similar to the existing "examples" folder?

File: test/http_memprof/http_memprof.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+var (
+	grpcAddr          = flag.String("grpc_addr", "127.0.0.1:12345", "Address of grpc server.")
+	profAddr          = flag.String("prof_addr", "http://localhost:6060", "Address of HTTP profile server.")
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:40]: use 127.0.0.1 for consistency?

File: test/http_memprof/http_memprof.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	outFolderParent   = flag.String("out_folder", "http_memprof_out", "Folder to store output of memory profiles over HTTP")
+	resultCsvFilename = "results.csv"
+)
+
+func readProfile(filename string) (*profile.Profile, error) {
+	reader, err := os.Open(filename)
+	if err != nil {
+		return nil, errors.New("file could not be opened: " + filename)
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:48]: should also long the original err here, or simply "return nil, err"

File: test/http_memprof/http_memprof.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	prof, err := profile.Parse(reader)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse proto file: %v", filename)
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:53]: consistently use errors.New or fmt.Errorf. Or, can just "return nil, err"

File: test/http_memprof/http_memprof.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	}
+	return prof, err
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:55]: "return prof, nil"

File: test/http_memprof/http_memprof.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+			if sample.DiffBaseSample() {
+				diffTotal += v
+			}
+		}
+		if diffTotal > 0 {
+			total = diffTotal
+		}
+		return total
+	}
+	// sample index of alloc_space in neap profiles from http/net/pprof
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:85]: neap -> heap?

File: test/http_memprof/http_memprof.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	const allocSpaceSampleIndex = 1
+
+	prof, err := readProfile(filename)
+	if err != nil {
+		log.Fatal(err)
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:90]: you can let the function return (int64, err) and just do "return 0, err" here.  In general, golang passes the err out and only let the outmost function to handle.

File: test/http_memprof/http_memprof.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+func saveProfile(outFolder string) {
+	outputFlag := fmt.Sprintf("-output=%v", outFolder)
+	heapProfileHTTPPath := fmt.Sprintf("%v/debug/pprof/heap?gc=1", *profAddr)
+	cmd := exec.Command("go", "tool", "pprof", outputFlag, "-proto", heapProfileHTTPPath)
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		fmt.Println(fmt.Sprint(err) + ": " + stderr.String())
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:111]: here too, can just do "return cmd.Run()" and let the function return error

File: test/http_memprof/http_memprof.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+	fmt.Printf("Attempting to make a connection to the gRPC server at %v\n", *grpcAddr)
+	conn, err := grpc.Dial(*grpcAddr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(100000000 /* 100M */)),
+		// The grpc.WithBlock() option make this call block until a connection is
+		// established, internally retrying until needed.
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		log.Printf("Could not connect: %v\n", err)
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:208]: log.Fatalf here

File: test/http_memprof/http_memprof.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+	log.Println("Connected to gRPC succesfully")
+
+	defer conn.Close()
+	c := pb.NewMixerClient(conn)
+	ctx := context.Background()
+
+	var profileResults []*MemoryProfileResult
+	var count int64
+	for _, allFacets := range []bool{true, false} {
+		for _, r := range []struct {
```
Comment (shifucun) [inline test/http_memprof/http_memprof.go:219]: Can you add a TODO here to move these into a config based file to incorporate for more APIs.

File: cmd/main.go [hunk 11]
```diff
@@ -236,6 +240,17 @@ func main() {
...
 		return;
 	}
+
+	// Launch a goroutine that will serve memory requests using net/http/pprof
+	if *httpProfilePort != 0 {
+		go func(){
+			// Code from https://pkg.go.dev/net/http/pprof README
+			httpProfileFrom := fmt.Sprintf("localhost:%d", *httpProfilePort)
+			log.Printf("Serving profile over HTTP on %v", httpProfileFrom)
+			log.Println(http.ListenAndServe(httpProfileFrom, nil))
```
[the remark this answers is not in the document]
Comment (enjoythecode) [reply cmd/main.go:250]: ListenAndServe is blocking, and it only returns non-nill errors [(documentation)](https://pantheon.corp.google.com/cloud-build/builds;region=global/2c36db30-8e14-40a0-86eb-30bb57c716f1?mods=-monitoring_api_staging&project=datcom-ci).

[the remark this answers is not in the document]
Comment (shifucun) [reply cmd/main.go:250]: Can follow patterns like https://source.corp.google.com/search?q=http.ListenAndServe%20lang:go&sq=

[the remark this answers is not in the document]
Comment (enjoythecode) [reply cmd/main.go:250]: changed to:
```golang
httpProfileFrom := fmt.Sprintf("localhost:%d", *httpProfilePort)   
log.Printf("Serving profile over HTTP on %v", httpProfileFrom)     
log.Printf("%s\n", http.ListenAndServe(httpProfileFrom, nil))
```

File: test/http_memprof/http_memprof.go (hunk 1 above)
Comment (enjoythecode) [reply test/http_memprof/http_memprof.go:17]: Yes, this is similar to "examples", but this code has functionality so I did not think that "examples" was the correct location. I will try again if the "deadcode" lint is still raised, however, I recall main() being detected as deadcode, which meant all of the file was deadcode since it is not called from anywhere else.

Comment (enjoythecode) [reply test/http_memprof/http_memprof.go]: Test passed; removing nolint:deadcode still works

File: test/test_driver.go [hunk 12]
```diff
@@ -11,6 +11,7 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+// noop
```
Comment (shifucun) [inline test/test_driver.go:14]: is this change needed?

**Review by shifucun — APPROVED** [review]
